File: dropdeck/slot.py
"""One slot on the board, and how it describes itself to a screen reader."""

from __future__ import annotations

import os
import random
from dataclasses import dataclass, field

from . import constants as C


def bank_for_index(index: int) -> int:
    """1-based bank number for a 0-based global slot index."""
    return index // C.SLOTS_PER_BANK + 1


def position_in_bank(index: int) -> int:
    """1-based position of a slot within its own bank."""
    return index % C.SLOTS_PER_BANK + 1


def _maybe_int(value):
    """An int off disk, or None. A folder count is derived and never trusted."""
    try:
        return max(0, int(value))
    except (TypeError, ValueError):
        return None


def format_duration(seconds: float | None) -> str:
    """A spoken-friendly duration. No bare colons for a screen reader to trip on."""
    if not seconds or seconds <= 0:
        return ""
    seconds = int(round(seconds))
    minutes, secs = divmod(seconds, 60)
    if minutes and secs:
        return f"{minutes} min {secs} sec"
    if minutes:
        return f"{minutes} min"
    return f"{secs} sec"


@dataclass
class Slot:
    """A single button's worth of state.

    ``index`` is the 0-based global position, 0 to 79. Everything else is either
    what the user assigned or what we measured when they assigned it.
    """

    #: Bank number -> the name the user gave that bank, shared with the Board
    #: that owns this slot. Deliberately NOT a dataclass field: it is the
    #: board's state rather than the slot's, it must never reach to_dict, and a
    #: slot built on its own in a test falls back to the shipped names.
    bank_names = None

    index: int
    filepath: str | None = None
    name: str | None = None
    duration: float | None = None
    custom_hotkey: str | None = None
    key_code: int | None = None
    modifiers: int | None = None
    #: A system-wide hotkey, which fires this slot while another window has
    #: focus. Separate from custom_hotkey, which only works inside the app.
    #: Always needs a modifier - see globalhotkeys.parse.
    global_hotkey: str | None = None
    loop: bool = field(default=False)
    #: Per-slot trim in decibels, so one loud sound can be tamed on its own.
    trim_db: float = 0.0
    #: How many playable sounds are in this slot's folder, when ``filepath`` is
    #: a folder rather than a file. Saved so the label is right before the first
    #: scan, which then corrects it. ``None`` for an ordinary slot.
    folder_count: int | None = None

    def __post_init__(self):
        #: The folder's contents, scanned. Never saved: it is what is on disk
        #: right now, and the point of a folder slot is that you can drop
        #: another jingle in without touching the app.
        self._folder_files: list = []
        self._folder_stamp = None
        self._last_pick = None

    # ------------------------------------------------------------- identity --
    @property
    def bank(self) -> int:
        return bank_for_index(self.index)

    @property
    def bank_title(self) -> str:
        """What this slot's bank is called - the user's name for it, if any."""
        return (self.bank_names or {}).get(self.bank) or C.BANK_TITLES[self.bank]

    @property
    def bank_short(self) -> str:
        """The same, short, for lists that span every bank.

        A renamed bank has no short form and inventing one would be guesswork,
        so the name the user typed is used whole. "SFX" is only a contraction
        of a name they did not choose.
        """
        return (self.bank_names or {}).get(self.bank) or C.BANK_SHORT[self.bank]

    @property
    def number(self) -> int:
        return position_in_bank(self.index)

    @property
    def is_bed(self) -> bool:
        return self.bank == C.LOOPING_BANK

    @property
    def hotkey_label(self) -> str:
        """The key that fires this slot, fixed by bank or chosen by the user."""
        if self.bank == C.BANK_MISC:
            return self.custom_hotkey or ""
        return C.BANK_HOTKEY_LABELS[self.bank][self.number - 1]

    # -------------------------------------------------------------- content --
    @property
    def is_assigned(self) -> bool:
        return bool(self.filepath)

    @property
    def is_folder(self) -> bool:
        """This slot holds a folder, and plays a different sound every press.

        Brian Hartgen's: a chart countdown has half a dozen "down the chart"
        jingles and you do not care which one you get, only that one plays. So
        the slot points at the folder and picks for you.
        """
        return bool(self.filepath) and os.path.isdir(self.filepath)

    @property
    def is_missing(self) -> bool:
        """Assigned, but the file is not where it used to be."""
        return bool(self.filepath) and not os.path.exists(self.filepath)

    @property
    def display_name(self) -> str:
        if self.name:
            return self.name
        if self.filepath:
            return os.path.splitext(os.path.basename(self.filepath))[0]
        return "Empty"

    def _state_words(self, playing: bool) -> list:
        state = []
        if self.is_missing:
            state.append("folder missing" if self.folder_count is not None
                         else "file missing")
        elif self.is_assigned:
            if playing:
                state.append("playing")
            if self.loop:
                state.append("loops")
            if self.is_folder:
                # A count, not a length. Every press is a different file, so a
                # duration here would be a different lie each time.
                n = self.folder_count
                state.append("folder, " + ("empty" if n == 0 else
                                           "1 sound" if n == 1 else
                                           "%d sounds" % n if n else
                                           "not counted yet"))
            else:
                dur = format_duration(self.duration)
                if dur:
                    state.append(dur)
        return state

    def button_label(self, playing: bool = False) -> str:
        """What the button says, and therefore what a screen reader reads.

        Name first, because that is what you are hunting for when you arrow
        along a row. The bank is not repeated here, the tab already said it.
        """
        parts = [f"{self.number}. {self.display_name}"]
        if self.hotkey_label:
            parts.append(f"key {self.hotkey_label}")
        if self.global_hotkey:
            parts.append(f"global {self.global_hotkey}")
        parts.extend(self._state_words(playing))
        return ", ".join(parts)

    def search_label(self, playing: bool = False) -> str:
        """Same idea, but for lists that span every bank, so name the bank."""
        parts = [self.display_name, f"{self.bank_short} {self.number}"]
        if self.hotkey_label:
            parts.append(f"key {self.hotkey_label}")
        if self.global_hotkey:
            parts.append(f"global {self.global_hotkey}")
        parts.extend(self._state_words(playing))
        return ", ".join(parts)

    # -------------------------------------------------------------- folders --
    def scan_folder(self, force: bool = False) -> int:
        """Re-read the folder if it has changed. Returns how many sounds it holds.

        Cheap to call often: it compares the folder's own timestamp and does
        nothing when nobody has touched it. It is never called between a
        keypress and a sound, the trigger path uses whatever the last scan
        found, and the cache warmer does the scanning at startup.
        """
        if not self.is_folder:
            self._folder_files = []
            return 0
        try:
            stamp = os.stat(self.filepath).st_mtime_ns
        except OSError:
            return len(self._folder_files)
        if not force and stamp == self._folder_stamp and self._folder_files:
            return len(self._folder_files)
        try:
            names = sorted(os.listdir(self.filepath))
        except OSError:
            return len(self._folder_files)
        self._folder_files = [
            os.path.join(self.filepath, name) for name in names
            if os.path.splitext(name)[1].lower() in C.AUDIO_EXTENSIONS
            and os.path.isfile(os.path.join(self.filepath, name))]
        self._folder_stamp = stamp
        self.folder_count = len(self._folder_files)
        return self.folder_count

    @property
    def folder_files(self) -> list:
        return list(self._folder_files)

    def pick_file(self) -> str | None:
        """One file out of the folder, at random, avoiding an instant repeat.

        Two presses in a row giving the same jingle is what makes a random
        stinger sound broken rather than random, so the last one is excluded
        whenever there is anything else to choose from.
        """
        if not self._folder_files:
            return None
        choices = self._folder_files
        if len(choices) > 1 and self._last_pick in choices:
            choices = [p for p in choices if p != self._last_pick]
        self._last_pick = random.choice(choices)
        return self._last_pick

    def playable_path(self) -> str | None:
        """What a press should play. A folder picks one; a file is itself."""
        if self.is_folder:
            return self.pick_file()
        return self.filepath

    # ----------------------------------------------------------- conversion --
    def to_dict(self) -> dict:
        return {
            "filepath": self.filepath,
            "name": self.name,
            "duration": self.duration,
            "custom_hotkey": self.custom_hotkey,
            "global_hotkey": self.global_hotkey,
            "key_code": self.key_code,
            "modifiers": self.modifiers,
            "loop": bool(self.loop),
            "trim_db": float(self.trim_db),
            "folder_count": self.folder_count,
        }

    @classmethod
    def from_dict(cls, index: int, data: dict | None) -> "Slot":
        data = data or {}
        return cls(
            index=index,
            filepath=data.get("filepath") or None,
            name=data.get("name") or None,
            duration=data.get("duration"),
            custom_hotkey=data.get("custom_hotkey") or None,
            global_hotkey=data.get("global_hotkey") or None,
            key_code=data.get("key_code"),
            modifiers=data.get("modifiers"),
            loop=bool(data.get("loop", bank_for_index(index) == C.LOOPING_BANK)),
            trim_db=float(data.get("trim_db", 0.0) or 0.0),
            folder_count=_maybe_int(data.get("folder_count")),
        )

    def clear(self) -> None:
        """Empty the slot but keep any custom hotkey the user set up."""
        self.filepath = None
        self.name = None
        self.duration = None
        self.trim_db = 0.0
        self.folder_count = None
        self._folder_files = []
        self._folder_stamp = None
        self._last_pick = None
        self.loop = self.is_bed

File: mac/Sources/Slot.swift
// One slot on the board, and how it describes itself to a screen reader.
//
// A faithful port of dropdeck/slot.py. The label composition in particular is
// not to be improved on: the order of the words in button_label is what a
// screen reader reads when you arrow along a row, and it was arrived at by
// listening to it.

import Foundation

func bankForIndex(_ index: Int) -> Int { index / C.slotsPerBank + 1 }
func positionInBank(_ index: Int) -> Int { index % C.slotsPerBank + 1 }

/// A spoken friendly duration. No bare colons for a screen reader to trip on.
func formatDuration(_ seconds: Double?) -> String {
    guard let seconds, seconds > 0 else { return "" }
    let total = Int(seconds.rounded())
    let minutes = total / 60
    let secs = total % 60
    if minutes > 0 && secs > 0 { return "\(minutes) min \(secs) sec" }
    if minutes > 0 { return "\(minutes) min" }
    return "\(secs) sec"
}

final class Slot {

    /// Bank number to the name the user gave that bank, shared by reference
    /// with the Board that owns this slot. Deliberately not part of the slot's
    /// own saved state: it is the board's, it must never reach toDict, and a
    /// slot built on its own in a test falls back to the shipped names.
    weak var board: Board?

    let index: Int
    var filepath: String?
    var name: String?
    var duration: Double?
    var customHotkey: String?
    /// The Windows key code and modifier mask. Carried through untouched so a
    /// board written here still fires the right key when it is opened on the
    /// Windows copy. This build never interprets them.
    var winKeyCode: Int?
    var winModifiers: Int?
    /// The Mac equivalent, which is what this build actually uses. A separate
    /// pair rather than a reused one, so neither platform can quietly destroy
    /// the other's binding.
    var macKeyCode: Int?
    var macModifiers: UInt?
    /// A system wide hotkey, which fires this slot while another window has
    /// focus. Separate from customHotkey, which only works inside the app.
    var globalHotkey: String?
    var loop: Bool
    /// Per slot trim in decibels, so one loud sound can be tamed on its own.
    var trimDB: Double = 0.0
    /// Taken off the board. The slot keeps everything it had, including its
    /// sound and its hotkeys, and can be put back; it simply is not shown and
    /// its key does nothing. Removing one NEVER renumbers the others, because
    /// the digit map is muscle memory: take slot 5 away and 6 is still on the
    /// 6 key.
    var hidden: Bool = false
    /// Pressing this slot's key again stops it, instead of playing a second
    /// copy on top of the first.
    ///
    /// Off by default, and deliberately. Effects and drops overlapping is the
    /// thing a soundboard is FOR: a laugh landing on top of a sting is the
    /// point, and making every key a toggle would take that away from everyone
    /// to help the one slot that needed it. Beds have always toggled and go on
    /// doing so whatever this says.
    var toggleStop: Bool = false
    /// How many playable sounds are in this slot's folder, when filepath is a
    /// folder rather than a file. Saved so the label is right before the first
    /// scan, which then corrects it. nil for an ordinary slot.
    var folderCount: Int?

    /// Anything in the saved slot this build did not recognise, kept so a round
    /// trip through the Mac copy cannot lose a Windows only field.
    var unknown: [String: Any] = [:]

    // The folder's contents, scanned. Never saved: it is what is on disk right
    // now, and the point of a folder slot is that you can drop another jingle
    // in without touching the app.
    private var folderFiles: [String] = []
    private var folderStamp: Int?
    private var lastPick: String?

    init(index: Int) {
        self.index = index
        self.loop = bankForIndex(index) == C.loopingBank
    }

    // ----------------------------------------------------------- identity ---
    var bank: Int { bankForIndex(index) }
    var number: Int { positionInBank(index) }
    var isBed: Bool { bank == C.loopingBank }

    /// What this slot's bank is called, the user's name for it if any.
    var bankTitle: String {
        board?.bankNames[bank] ?? C.bankTitles[bank] ?? ""
    }

    /// The same, short, for lists that span every bank.
    ///
    /// A renamed bank has no short form and inventing one would be guesswork,
    /// so the name the user typed is used whole. "SFX" is only a contraction of
    /// a name they did not choose.
    var bankShort: String {
        board?.bankNames[bank] ?? C.bankShort[bank] ?? ""
    }

    /// The key that fires this slot, fixed by bank or chosen by the user.
    var hotkeyLabel: String {
        if bank == C.bankMisc { return customHotkey ?? "" }
        return KeyMap.hotkeyLabel(bank: bank, positionInBank: number - 1)
    }

    // ------------------------------------------------------------ content ---
    var isAssigned: Bool { !(filepath ?? "").isEmpty }

    /// This slot holds a folder, and plays a different sound every press.
    ///
    /// A chart countdown has half a dozen "down the chart" jingles and you do
    /// not care which one you get, only that one plays. So the slot points at
    /// the folder and picks for you.
    var isFolder: Bool {
        guard let p = filepath, !p.isEmpty else { return false }
        var isDir: ObjCBool = false
        return FileManager.default.fileExists(atPath: p, isDirectory: &isDir) && isDir.boolValue
    }

    /// Assigned, but the file is not where it used to be.
    var isMissing: Bool {
        guard let p = filepath, !p.isEmpty else { return false }
        return !FileManager.default.fileExists(atPath: p)
    }

    var displayName: String {
        if let n = name, !n.isEmpty { return n }
        if let p = filepath, !p.isEmpty {
            return (p as NSString).lastPathComponent.replacingOccurrences(
                of: "." + (p as NSString).pathExtension, with: "")
        }
        return "Empty"
    }

    private func stateWords(playing: Bool) -> [String] {
        var state: [String] = []
        if isMissing {
            state.append(folderCount != nil ? "folder missing" : "file missing")
        } else if isAssigned {
            if playing { state.append("playing") }
            if loop { state.append("loops") }
            if isFolder {
                // A count, not a length. Every press is a different file, so a
                // duration here would be a different lie each time.
                if let n = folderCount {
                    state.append("folder, " + (n == 0 ? "empty"
                                               : n == 1 ? "1 sound"
                                               : "\(n) sounds"))
                } else {
                    state.append("folder, not counted yet")
                }
            } else {
                let dur = formatDuration(duration)
                if !dur.isEmpty { state.append(dur) }
            }
        }
        return state
    }

    /// What the button says, and therefore what a screen reader reads.
    ///
    /// Name first, because that is what you are hunting for when you arrow
    /// along a row. The bank is not repeated here, the tab already said it.
    func buttonLabel(playing: Bool = false) -> String {
        var parts = ["\(number). \(displayName)"]
        if !hotkeyLabel.isEmpty { parts.append("key \(hotkeyLabel)") }
        if let g = globalHotkey, !g.isEmpty { parts.append("global \(g)") }
        parts.append(contentsOf: stateWords(playing: playing))
        return parts.joined(separator: ", ")
    }

    /// Same idea, but for lists that span every bank, so name the bank.
    func searchLabel(playing: Bool = false) -> String {
        var parts = [displayName, "\(bankShort) \(number)"]
        if !hotkeyLabel.isEmpty { parts.append("key \(hotkeyLabel)") }
        if let g = globalHotkey, !g.isEmpty { parts.append("global \(g)") }
        parts.append(contentsOf: stateWords(playing: playing))
        return parts.joined(separator: ", ")
    }

    // ------------------------------------------------------------ folders ---

    /// Re-read the folder if it has changed. Returns how many sounds it holds.
    ///
    /// Cheap to call often: it compares the folder's own timestamp and does
    /// nothing when nobody has touched it. It is never called between a
    /// keypress and a sound, the trigger path uses whatever the last scan
    /// found, and the cache warmer does the scanning at startup.
    @discardableResult
    func scanFolder(force: Bool = false) -> Int {
        guard isFolder, let path = filepath else {
            folderFiles = []
            return 0
        }
        let fm = FileManager.default
        guard let attrs = try? fm.attributesOfItem(atPath: path),
              let modified = attrs[.modificationDate] as? Date else {
            return folderFiles.count
        }
        let stamp = Int(modified.timeIntervalSince1970 * 1_000_000_000)
        if !force, stamp == folderStamp, !folderFiles.isEmpty {
            return folderFiles.count
        }
        guard let names = try? fm.contentsOfDirectory(atPath: path) else {
            return folderFiles.count
        }
        folderFiles = names.sorted().compactMap { name in
            let full = (path as NSString).appendingPathComponent(name)
            let ext = (name as NSString).pathExtension.lowercased()
            guard AudioFile.supportedExtensions.contains(ext) else { return nil }
            var isDir: ObjCBool = false
            guard fm.fileExists(atPath: full, isDirectory: &isDir), !isDir.boolValue
            else { return nil }
            return full
        }
        folderStamp = stamp
        folderCount = folderFiles.count
        return folderFiles.count
    }

    var folderFileList: [String] { folderFiles }

    /// One file out of the folder, at random, avoiding an instant repeat.
    ///
    /// Two presses in a row giving the same jingle is what makes a random
    /// stinger sound broken rather than random, so the last one is excluded
    /// whenever there is anything else to choose from.
    func pickFile() -> String? {
        guard !folderFiles.isEmpty else { return nil }
        var choices = folderFiles
        if choices.count > 1, let last = lastPick, choices.contains(last) {
            choices = choices.filter { $0 != last }
        }
        lastPick = choices.randomElement()
        return lastPick
    }

    /// What a press should play. A folder picks one; a file is itself.
    func playablePath() -> String? {
        isFolder ? pickFile() : filepath
    }

    /// Empty the slot but keep any custom hotkey the user set up.
    func clear() {
        filepath = nil
        name = nil
        duration = nil
        trimDB = 0.0
        folderCount = nil
        folderFiles = []
        folderStamp = nil
        lastPick = nil
        loop = isBed
    }

    // --------------------------------------------------------- conversion ---
    func toDict() -> [String: Any] {
        var d = unknown
        d["filepath"] = filepath as Any? ?? NSNull()
        d["name"] = name as Any? ?? NSNull()
        d["duration"] = duration as Any? ?? NSNull()
        d["custom_hotkey"] = customHotkey as Any? ?? NSNull()
        d["global_hotkey"] = globalHotkey as Any? ?? NSNull()
        d["key_code"] = winKeyCode as Any? ?? NSNull()
        d["modifiers"] = winModifiers as Any? ?? NSNull()
        d["mac_key_code"] = macKeyCode as Any? ?? NSNull()
        d["mac_modifiers"] = macModifiers as Any? ?? NSNull()
        d["loop"] = loop
        d["trim_db"] = trimDB
        d["folder_count"] = folderCount as Any? ?? NSNull()
        d["hidden"] = hidden
        d["toggle_stop"] = toggleStop
        return d
    }

    static func fromDict(index: Int, data: [String: Any]?) -> Slot {
        let s = Slot(index: index)
        guard let data else { return s }
        let known: Set<String> = ["filepath", "name", "duration", "custom_hotkey",
                                  "global_hotkey", "key_code", "modifiers",
                                  "mac_key_code", "mac_modifiers", "loop",
                                  "trim_db", "folder_count", "hidden", "toggle_stop"]
        s.unknown = data.filter { !known.contains($0.key) }

        s.filepath = data["filepath"] as? String
        s.name = data["name"] as? String
        s.duration = data["duration"] as? Double
        s.customHotkey = data["custom_hotkey"] as? String
        s.globalHotkey = data["global_hotkey"] as? String
        s.winKeyCode = data["key_code"] as? Int
        s.winModifiers = data["modifiers"] as? Int
        s.macKeyCode = data["mac_key_code"] as? Int
        s.macModifiers = data["mac_modifiers"] as? UInt
        s.loop = data["loop"] as? Bool ?? (bankForIndex(index) == C.loopingBank)
        s.trimDB = data["trim_db"] as? Double ?? 0.0
        if let fc = data["folder_count"] as? Int { s.folderCount = max(0, fc) }
        s.hidden = data["hidden"] as? Bool ?? false
        s.toggleStop = data["toggle_stop"] as? Bool ?? false
        return s
    }
}
